executor/src/lib.rs

```diff
@@ -752,72 +752,91 @@ where
         let mut next_validator_set = None;
 
         let proof_reader = ProofReader::new(account_to_proof);
+        let mut validator_set_change_seen = false;
         for (vm_output, txn) in itertools::zip_eq(vm_outputs.into_iter(), transactions.iter()) {
-            let (blobs, state_tree, num_accounts_created) = Self::process_write_set(
-                txn,
-                &mut account_to_state,
-                &proof_reader,
-                vm_output.write_set().clone(),
-                &current_state_tree,
-            )?;
-
-            let event_tree = {
-                let event_hashes: Vec<_> =
-                    vm_output.events().iter().map(CryptoHash::hash).collect();
-                InMemoryAccumulator::<EventAccumulatorHasher>::from_leaves(&event_hashes)
-            };
-            let mut txn_info_hash = None;
-
-            match vm_output.status() {
-                TransactionStatus::Keep(status) => {
-                    ensure!(
-                        !vm_output.write_set().is_empty(),
-                        "Transaction with empty write set should be discarded.",
-                    );
-                    // Compute hash for the TransactionInfo object. We need the hash of the
-                    // transaction itself, the state root hash as well as the event root hash.
-                    let txn_info = TransactionInfo::new(
-                        txn.hash(),
-                        state_tree.root_hash(),
-                        event_tree.root_hash(),
-                        vm_output.gas_used(),
-                        status.major_status,
-                    );
-
-                    let real_txn_info_hash = txn_info.hash();
-                    txn_info_hashes.push(real_txn_info_hash);
-                    txn_info_hash = Some(real_txn_info_hash);
-                }
-                TransactionStatus::Discard(status) => {
-                    if !vm_output.write_set().is_empty() || !vm_output.events().is_empty() {
-                        crit!(
-                            "Discarded transaction has non-empty write set or events. \
-                             Transaction: {:?}. Status: {}.",
-                            txn,
-                            status,
+            if validator_set_change_seen {
+                txn_data.push(TransactionData::new(
+                    HashMap::new(),
+                    vec![],
+                    TransactionStatus::Retry,
+                    Arc::clone(&current_state_tree),
+                    Arc::new(InMemoryAccumulator::<EventAccumulatorHasher>::default()),
+                    0,
+                    0,
+                    None,
+                ));
+            } else {
```

Comment:
recommend using a continue statement so you don't need to add this else and indent the entire huge chunk of code below making it easier to read.